packages/substyle/test/useStyles.spec.js

```diff
@@ -0,0 +1,170 @@
+      defaultContainerStyle,
+      { className, style },
+      { '&readOnly': readOnly }
+    )
+    return (
+      <div {...styles}>
+        <Content style={styles('content')} />
+
+        <div {...styles('footer')} />
+      </div>
+    )
+  }
+
+  it('should correctly apply default styles and derive class names', () => {
+    const wrapper = mount(<Container className="foo" />)
+
+    expect(wrapper.find('div').at(0).prop('style')).toEqual({
+      background: 'white',
+    })
+
+    expect(wrapper.find('.foo__footer')).toExist()
+    expect(wrapper.find('.foo__footer').prop('style')).toEqual({
+      fontSize: 10,
+    })
+
+    expect(wrapper.find('.foo__content')).toExist()
+    expect(wrapper.find('.foo__content__title')).toExist()
+    expect(wrapper.find('.foo__content__title').prop('style')).toEqual({
+      fontStyle: 'bold',
+    })
+  })
+
+  it('should merge styles provided by the user with default styles', () => {
+    const wrapper = mount(<Container style={{ cursor: 'pointer' }} />)
+
+    expect(wrapper.find('div').at(0).prop('style')).toEqual({
+      background: 'white',
+      cursor: 'pointer',
+    })
+  })
+
+  it('should also work if a substyle instance is provided as `style`', () => {
+    const wrapper = mount(
+      <Container
+        style={createSubstyle({
+          style: { cursor: 'pointer' },
+        })}
+      />
+    )
+
+    expect(wrapper.find('div').at(0).prop('style')).toEqual({
+      background: 'white',
+      cursor: 'pointer',
+    })
+  })
+
+  it('should take a modifier selection function as second argument', () => {
+    const wrapper = mount(<Container readOnly />)
+    expect(wrapper.find('div').at(0).prop('style')).toHaveProperty(
+      'color',
+      'gray'
+    )
+  })
+
+  it('should apply the selected modifiers also on the style supplied by the user', () => {
+    const wrapper = mount(
+      <Container
+        readOnly
+        style={{
+          '&readOnly': {
+            opacity: 0.5,
+          },
+        }}
+      />
+    )
+
+    expect(wrapper.find('div').at(0).prop('style')).toHaveProperty(
+      'opacity',
+      0.5
+    )
+  })
+
+  it('should give precedence to styles supplied by the user, regardless the modifiers specificity', () => {
+    const wrapper = mount(
+      <Container
+        readOnly
+        style={{
+          opacity: 0.7,
+        }}
+      />
+    )
+    expect(wrapper.find('div').at(0).prop('style')).toHaveProperty(
+      'opacity',
+      0.7
+    )
+  })
+
+  it('should preserve styles if modifiers are shallow equal', () => {
+    let style
+    let defaultStyle
+
+    const Component = ({ readOnly }) => {
+      style = useStyles(defaultStyle, { '&readOnly': readOnly })
+      return null
+    }
+    const wrapper = mount(<Component />)
+    const styleOnFirstRender = style
+
+    wrapper.setProps({ update: 'yes' })
+
+    const styleOnSecondRender = style
+
+    expect(styleOnFirstRender).toBe(styleOnSecondRender)
+  })
```

Comment:
That's the test I was talking about. I thought this would test that memoization already works as expected and we don't need to do extra work. You're right that modifiers change on every render. Maybe the `useMemo` is then just not needed.